R/utils.R

```diff
@@ -0,0 +1,189 @@
+#' Sample names using [randomNames::randomNames()]
+#'
+#' @description
+#' Sample names for specified sexes by sampling with replacement to avoid
+#' exhausting number of name when `sample.with.replacement = FALSE`. The
+#' duplicated names during sampling need to be removed to ensure each
+#' individual has a unique name. In order to have enough unique names, more
+#' names than required are sampled from [randomNames()], and the level of
+#' oversampling is determined by the `buffer_factor` argument. A
+#' `buffer_factor` too high and the more names are sampled which takes longer,
+#' a `buffer_factor` too low and not enough unique names are sampled and
+#' the `.sample_names()` function will need to loop until it has enough
+#' unique names.
+#'
+#' @inheritParams .add_date
+#' @param buffer_factor A single `numeric` determining the level of
+#' oversampling (or buffer) when creating a vector of unique names from
+#' [randomNames()].
+#'
+#' @return A `character` vector.
+#' @keywords internal
+.sample_names <- function(.data,
+                          buffer_factor = 1.5) {
+  m_lgl_idx <- .data$sex == "m"
+  f_lgl_idx <- .data$sex == "f"
+  num_m <- sum(m_lgl_idx)
+  num_f <- sum(f_lgl_idx)
+  num_sample_m <- ceiling(num_m * buffer_factor)
+  num_sample_f <- ceiling(num_f * buffer_factor)
+
+  # create sample of names so there are no duplicates
+  names_m <- character(0)
+  while (length(names_m) < num_m) {
+    names_m <- unique(
+      randomNames::randomNames(
+        which.names = "both",
+        name.sep = " ",
+        name.order = "first.last",
+        gender = rep("M", num_sample_m),
+        sample.with.replacement = TRUE
+      )
+    )
+  }
+
+  names_f <- character(0)
+  while (length(names_f) < num_f) {
+    names_f <- unique(
+      randomNames::randomNames(
+        which.names = "both",
+        name.sep = " ",
+        name.order = "first.last",
+        gender = rep("F", num_sample_f),
+        sample.with.replacement = TRUE
+      )
+    )
+  }
+
+  # subset to use required number of names
+  names_m <- names_m[1:num_m]
+  names_f <- names_f[1:num_f]
+
+  # order names with sex codes from .data
+  names_mf <- vector(mode = "character", length = nrow(.data))
+  names_mf[m_lgl_idx] <- names_m
+  names_mf[f_lgl_idx] <- names_f
+
+  # return vector of names
+  names_mf
+}
+
+#' Anonymise names
+#'
+#' @description
+#' A simple algorithm to replace names with an alphanumeric string with an
+#' fixed number of characters (i.e. [nchar()]) specified by `string_len`.
+#'
+#' @param x A vector of `character` strings.
+#' @param string_len A single `numeric` specifying the number of alphanumeric
+#' characters to use for each anonymising `character` string.
+#' Default is `10`.
+#'
+#' @return A vector of `character` strings of equal length to the input.
+#' @keywords internal
+.anonymise <- function(x, string_len = 10) {
```

Comment:
I sometimes have a bit of friction with our minimal dependency requirement, as I have here - it seems like a lot of work to create a md5sum/shasum for strings, which does not seem possible in base R. It makes sense under minimal dependencies to create a custom function.

[**digest**](https://www.r-pkg.org/pkg/digest) provides this functionality as well and could save you these lines of code. It also seems quite widely used and may meet the availability expectations (although certainly not as available as **ggplot2**). 

Yet your implementation seems pretty good as well, so I am not going to argue for adding a dependency. Surfacing my thoughts on this so you can check-in on them if you'd like. Happy to discuss, but not a hard want 😊

Reply:
I did a quick search of packages that offer hashing and several looked appropriate. I decided to write an internal implementation as the package does not need all of the complexity offered by these packages and the complexity of the function was not that difficult/time consuming to write, and I don't think it adds much of a maintenance burden to the package. 

To me this kind of thing can go either way without much issue, either keep the internal function or import {digest} or equivalent package. I'll leave the current implementation for now as it seems to work fine, but we can reopen this discussion if things change. 😄